File: serve_web.py
#!/usr/bin/env python3
"""site-patrol 本地预览服务器。

解决"内置浏览器打开不是最新构建"的缓存问题：
1. 所有响应带 Cache-Control: no-store，禁止 HTTP 层缓存；
2. 拦截 Flutter 的 flutter_service_worker.js，返回"卸载脚本"——
   已注册的 Service Worker 会删除全部 CacheStorage 缓存并自杀，
   之后每次刷新都直接从磁盘读最新构建产物。

用法:
    python3 serve_web.py --port 8765 --directory build/web
"""
import argparse
import http.server
import socketserver

# 拦截 Service Worker：清除历史缓存并注销自身。
SERVICE_WORKER_KILL = b"""self.addEventListener('install', () => self.skipWaiting());
self.addEventListener('activate', (event) => {
  event.waitUntil((async () => {
    const keys = await caches.keys();
    await Promise.all(keys.map((k) => caches.delete(k)));
    await self.registration.unregister();
  })());
});
"""


class NoCacheHandler(http.server.SimpleHTTPRequestHandler):
    def _no_store_headers(self):
        self.send_header('Cache-Control', 'no-store, no-cache, must-revalidate, max-age=0')
        self.send_header('Pragma', 'no-cache')
        self.send_header('Expires', '0')

    def do_GET(self):
        # Flutter Service Worker：返回卸载脚本（带版本 query 也能命中）
        if self.path.split('?')[0].endswith('/flutter_service_worker.js'):
            self.send_response(200)
            self.send_header('Content-Type', 'application/javascript; charset=utf-8')
            self.send_header('Content-Length', str(len(SERVICE_WORKER_KILL)))
            self._no_store_headers()
            self.end_headers()
            self.wfile.write(SERVICE_WORKER_KILL)
            return
        super().do_GET()

    def end_headers(self):
        self._no_store_headers()
        super().end_headers()


def main():
    parser = argparse.ArgumentParser(description='No-cache static preview server')
    parser.add_argument('--port', type=int, default=8765)
    parser.add_argument('--host', default='0.0.0.0',
                        help='bind host; 0.0.0.0 允许局域网其他设备访问')
    parser.add_argument('--directory', default='.')
    args = parser.parse_args()

    handler = lambda *a, **kw: NoCacheHandler(*a, directory=args.directory, **kw)
    with socketserver.ThreadingTCPServer((args.host, args.port), handler) as httpd:
        print(f'Serving {os.path.abspath(args.directory)} on http://{args.host}:{args.port} (no-cache)')
        httpd.serve_forever()


if __name__ == '__main__':
    import os
    main()

File: _serve_web.js
// Zero-dependency static server for Flutter Web builds.
// Usage: node _serve_web.js [port] [rootDir]
// Features: correct MIME, Cache-Control: no-store (avoid stale builds),
//           SPA fallback to index.html, kill Flutter service worker cache.
'use strict';
const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = parseInt(process.argv[2], 10) || 8080;
const ROOT = path.resolve(__dirname, process.argv[3] || 'build/web');

const MIME = {
  '.html': 'text/html; charset=utf-8', '.htm': 'text/html; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8', '.mjs': 'application/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8', '.json': 'application/json; charset=utf-8',
  '.png': 'image/png', '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg',
  '.gif': 'image/gif', '.svg': 'image/svg+xml', '.ico': 'image/x-icon',
  '.webp': 'image/webp', '.wasm': 'application/wasm', '.ttf': 'font/ttf',
  '.woff': 'font/woff', '.woff2': 'font/woff2', '.otf': 'font/otf',
  '.txt': 'text/plain; charset=utf-8', '.map': 'application/json; charset=utf-8',
  '.pdf': 'application/pdf', '.ocf': 'application/octet-stream',
};

// Intercept service worker: purge old caches and unregister itself.
const SW_KILL = `self.addEventListener('install', () => self.skipWaiting());
self.addEventListener('activate', (event) => {
  event.waitUntil((async () => {
    const keys = await caches.keys();
    await Promise.all(keys.map((k) => caches.delete(k)));
    await self.registration.unregister();
  })());
});
`;

const NO_STORE = 'no-store, no-cache, must-revalidate, max-age=0';
const ASSET_RE = /\.(html?|js|mjs|css|json|png|jpe?g|gif|svg|ico|webp|wasm|ttf|woff2?|otf|txt|map|pdf|ocf)$/i;

http.createServer((req, res) => {
  let rel;
  try {
    rel = decodeURIComponent(new URL(req.url, 'http://localhost').pathname);
  } catch (e) {
    rel = req.url.split('?')[0];
  }

  // Flutter service worker -> uninstall script
  if (rel.endsWith('/flutter_service_worker.js')) {
    res.writeHead(200, {
      'Content-Type': 'application/javascript; charset=utf-8',
      'Content-Length': Buffer.byteLength(SW_KILL),
      'Cache-Control': NO_STORE,
    });
    res.end(SW_KILL);
    return;
  }

  if (rel === '/') rel = '/index.html';
  let file = path.normalize(path.join(ROOT, rel));
  if (!file.startsWith(ROOT)) { res.writeHead(403); res.end('403 Forbidden'); return; }

  const send = (code, ctype, data) => {
    res.writeHead(code, {
      'Content-Type': ctype,
      'Content-Length': data.length,
      'Cache-Control': NO_STORE,
    });
    res.end(data);
  };

  fs.stat(file, (err, st) => {
    if (!err && st.isDirectory()) file = path.join(file, 'index.html');
    fs.readFile(file, (err2, data) => {
      if (!err2) {
        const ext = path.extname(file).toLowerCase();
        send(200, MIME[ext] || 'application/octet-stream', data);
        return;
      }
      // SPA fallback only for non-asset paths
      if (!ASSET_RE.test(rel)) {
        fs.readFile(path.join(ROOT, 'index.html'), (err3, idx) => {
          if (err3) { send(404, 'text/plain; charset=utf-8', Buffer.from('404 Not Found')); return; }
          send(200, 'text/html; charset=utf-8', idx);
        });
        return;
      }
      send(404, 'text/plain; charset=utf-8', Buffer.from('404 Not Found: ' + rel));
    });
  });
}).listen(PORT, '0.0.0.0', () => {
  console.log(`serve http://localhost:${PORT}  root=${ROOT}  (no-cache)`);
});
